src/linker/Linker/Driver.cs

```diff
@@ -780,29 +781,41 @@ public int Run (ILogger? customLogger = null, bool throwOnFatalLinkerException =
 
 			try {
 				p.Process (Context);
-			} catch (LinkerFatalErrorException lex) {
+			} catch (Exception e) when (FailCatch (e)) {
+				// Proces will be terminated in exception filter.
+				throw e;
+			} finally {
+				Context.FlushCachedWarnings ();
+				Context.Tracer.Finish ();
+			}
+
+			return Context.ErrorsCount > 0 ? 1 : 0;
+		}
+
+		/// <summary>
+		/// Handles exceptions in the linker. Prints error messages, then calls Environment.FailFast to create a dump with the full call stack.
+		/// </summary>
+		bool FailCatch (Exception e)
+		{
+
+			switch (e) {
+			case LinkerFatalErrorException lex:
 				Context.LogMessage (lex.MessageContainer);
 				Console.Error.WriteLine (lex.ToString ());
 				Debug.Assert (lex.MessageContainer.Category == MessageCategory.Error);
 				Debug.Assert (lex.MessageContainer.Code != null);
 				Debug.Assert (lex.MessageContainer.Code.Value != 0);
-				if (throwOnFatalLinkerException)
-					throw;
-				return lex.MessageContainer.Code ?? 1;
-			} catch (ResolutionException e) {
-				Context.LogError (null, DiagnosticId.FailedToResolveMetadataElement, e.Message);
-			} catch (Exception) {
-				// Unhandled exceptions are usually linker bugs. Ask the user to report it.
+				break;
+			case ResolutionException re:
+				Context.LogError (null, DiagnosticId.FailedToResolveMetadataElement, re.Message);
+				break;
+			default:
 				Context.LogError (null, DiagnosticId.LinkerUnexpectedError);
-				// Don't swallow the exception and exit code - rethrow it and let the surrounding tooling decide what to do.
-				// The stack trace will go to stderr, and the MSBuild task will surface it with High importance.
-				throw;
-			} finally {
-				Context.FlushCachedWarnings ();
-				Context.Tracer.Finish ();
+				break;
 			}
-
-			return Context.ErrorsCount > 0 ? 1 : 0;
+			Context.FlushCachedWarnings ();
```

Comment:
I'm a bit nervous about this. Ideally we would do very little in this method since we can't really trust data structures anymore. I don't see why we would need to print out all the warnings we got so far on a fatal crash. I know that there's some chance that it might help, but in the worst case if we get a dump the warnings will be in the dump.

Same goes about the tracer - the tracer produces the large XML files which are rarely used and only to figure out "why was this kept". Not having that info complete when the linker hard crashes is absolutely fine.

Definitely worth a comment that we intentionally don't do it.

Reply:
We'll probably want to move them out of the `finally` block too then, right?

Reply:
Looking closer I realize messages and errors are not cached, so we would be able to log some extra information without flushing.